src/frames.ts

```diff
@@ -391,48 +396,78 @@ export class Frame {
       }
     };
 
-    let error = await Promise.race([
+    throwIfError(await Promise.race([
       navigate(),
-      watcher.timeoutOrTerminationPromise,
-    ]);
-    if (!error) {
-      const promises = [watcher.timeoutOrTerminationPromise];
-      if (navigateResult!.newDocumentId) {
-        watcher.setExpectedDocumentId(navigateResult!.newDocumentId, url);
-        promises.push(watcher.newDocumentNavigationPromise);
-      } else if (navigateResult!.isSameDocument) {
-        promises.push(watcher.sameDocumentNavigationPromise);
-      } else {
-        promises.push(watcher.sameDocumentNavigationPromise, watcher.newDocumentNavigationPromise);
-      }
-      error = await Promise.race(promises);
+      timeoutPromise,
+      frameDestroyedPromise,
+    ]));
+
+    const promises: Promise<Error|void>[] = [timeoutPromise, frameDestroyedPromise];
+    if (navigateResult!.newDocumentId)
+      promises.push(disposer.add(waitForSpecificDocument(this, navigateResult!.newDocumentId)));
+    else
+      promises.push(sameDocumentPromise);
+    throwIfError(await Promise.race(promises));
+
+    const request = (navigateResult! && navigateResult!.newDocumentId) ? requestWatcher.get(navigateResult!.newDocumentId) : null;
+    const waitForLifecyclePromise = disposer.add(waitForLifecycle(this, options.waitUntil));
+    throwIfError(await Promise.race([timeoutPromise, frameDestroyedPromise, waitForLifecyclePromise]));
+
+    disposer.dispose();
+
+    return request ? request!._finalRequest._waitForResponse() : null;
+
+    function throwIfError(error: Error|void): asserts error is void {
+      if (!error)
+        return;
+      disposer.dispose();
+      const message = `While navigating to ${url}: ${error.message}`;
+      if (error instanceof TimeoutError)
+        throw new TimeoutError(message);
+      throw new Error(message);
     }
-    watcher.dispose();
-    if (error)
-      throw error;
-    return watcher.navigationResponse();
   }
 
-  async waitForNavigation(options?: WaitForNavigationOptions): Promise<network.Response | null> {
-    const watcher = new LifecycleWatcher(this, options, true /* supportUrlMatch */);
-    const error = await Promise.race([
-      watcher.timeoutOrTerminationPromise,
-      watcher.sameDocumentNavigationPromise,
-      watcher.newDocumentNavigationPromise,
+  async waitForNavigation(options: WaitForNavigationOptions = {}): Promise<network.Response | null> {
+    const disposer = new Disposer();
+    const requestWatcher = disposer.add(trackDocumentRequests(this));
+    const {timeout = this._page._timeoutSettings.navigationTimeout()} = options;
+
+    const failurePromise = Promise.race([
+      disposer.add(createTimeoutPromise(timeout)),
+      disposer.add(createFrameDestroyedPromise(this)),
+    ]);
+    let documentId : string|null = null;
+    let error : void|Error = await Promise.race([
```

Comment:
nit: I see a lot of `foo : bar` with extra spacing, shouldn't eslint complain?

Reply:
#897